### ISSUE TITLE
Custom filter functions for graph traversals in version 3.2

### ISSUE THREAD

**Issue by felixniemeyer** [arangodb/arangodb#2858]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ ] 3.1
- [X] 3.2 Beta1
- [ ] self-compiled devel branch

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [X] own infrastructure
- [ ] Linux 
  - [ ] Debian .deb
  - [X] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [ ] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is an AQL-related issue:
- [X] I'm using graph features

I'm issuing AQL via:
- [X] web interface with this browser:     running on this OS:
- [X] arangosh
- [X] this Driver: anrangojs

The question I have is: has custom traversal filter function support been removed as advancing from version 2.8 to 3.2?

I read that "the optimizer can detect that it is possible to prune paths" in 3.2, but I don't trust it completely and think it would be cool to write custom traversal filter functions :+1:   

I tried to combine the 2.8 and 3.2 Syntax and it looks like this and does not work. It might illustrate what I aim to do though:

```
LET params = {
    filterVertices: "ns::someFunction"
}

FOR vertex, edge, path IN OUTBOUND "activities/root4313031" connections OPTIONS {filterVertices: "synachieve::accessInheritanceFilter"}
 RETURN vertex.title
```
yea, and the function itself was registered based on this V2.8 article https://docs.arangodb.com/2.8/cookbook/FilteringNodesGraph.html
and looks like this: 

```
'use strict'

module.exports = function(config, vertex, path) {
	if(vertex.someAttribute)
		return ["prune", "exclude"];
	else 
		return ["prune", "exclude"];
}

```

registered in arangosh with 
```
aqlfunctions.register("synachieve::noAccessInheritance", func, true)
```

**Comment by jsteemann** [comment]
@felixniemeyer AFAIK support for custom filter functions has been removed in 3.0, because the traversal code was completely rewritten internally for that particular release. 

@mchacki should confirm this however.

**Comment by sowee15** [comment]
I asked a [similar question on stackoverflow](https://stackoverflow.com/questions/44398294/traversal-flow-control-in-arangodb-3-0) some time ago, I am also interested in customizing the traversal logic in arangodb 3+ (for instance, only follow edges that have a certain attribute value).

**Comment by dothebart** [comment]
@sowee15 you should be able to do that [with pattern matching on paths](https://docs.arangodb.com/3.2/AQL/Graphs/Traversals.html#filtering-on-paths) 

You could filter for: `path.edges[*].member != null` - which will work if your attribute is either set by a non-null value or unset. However, documents having that attribute set will also be hit by this filter. 

[In the future the usage of `has()` may become a possibility](https://docs.arangodb.com/3.2/AQL/Functions/Document.html#has) - but right now its not.

**Comment by Simran-B** [comment]
We lack support for complex conditions on the path as far as I can tell (`ALL ==` can not be combined with an alternative `OR` condition). Such filter conditions appear to be possible with post-filtering, but not during traversal (pruning). I'll create an internal ticket about this limit in composability to gather the information currently spread over multiple information channels.

**Comment by sowee15** [comment]
@dothebart I think the old custom visitors gave the possibility to stop exploration of a certain branch, instead of just filtering results but still exploring them. If the graph is huge, this could incur some overhead?

**Comment by dothebart** [comment]
Yes, 
the old javascript (<= ArangoDB 2.8) would allow you to specify such a javascript function. 
Once you're working on a graph of reasonable size that would outdo the capabilities of the pattern matching traverser, you would have long left the capabilities of the old graph module behind. 

Since such a javascript function would have to be called frequently, and certain overhead is involved by that, we didn't implement it yet.

**Comment by sowee15** [comment]
So, right now, the best way to filter the resultset is to use what's available in AQL in the traversal operator (either filtering on the vertex, edge or path). That won't stop exploration though (maybe I'm over estimating the cost of exploration), is that right? Or would filtering on the path have the effect of stopping exploration of current path? Thanks for the clarifications.

**Comment by OmarAyo** [comment]
Hi there,

This should mostly be solvable by the PRUNE keyword now available starting from ArangoDB Version 3.5.0 see [documentation](https://www.arangodb.com/docs/devel/aql/graphs-traversals.html#pruning)

Closing,
